### PR TITLE
Document SAM registration (#2286)

### PR DESCRIPTION
**Pull request by amarjandu**
Author (before primary review)

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue
- [x] Added `HCA` label <sub>or this PR does not target an `hca/*` branch</sub>
- [x] Created issue to track porting these changes <sub>or this PR does not need porting</sub> 
- [x] Added `[r]` prefix at start of commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label <sub>or this PR does not require reindexing</sub>
- [x] Freebies are blocked on this PR <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Ran `make requirements_update` <sub>or this PR does not change requirements*.txt</sub>
- [x] Mentioned `make requirements` in UPGRADING.rst <sub>or this PR does not change requirements*.txt</sub>
- [x] Documented upgrade of personal deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>

Primary reviewer (before pushing merge commit)

- [x] Checked `reindex` label and `[r]` prefix of commit title
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Build passes in `sandbox` <sub>or commented that sandbox will be skipped</sub>
- [x] Reindexed `sandbox` <sub>or this PR does not require reindexing</sub>
- [x] Added PR reference to merge commit
- [x] Moved linked issue to Merged
- [ ] Pushed merge commit to Github

Primary reviewer (after pushing merge commit)

- [ ] Moved freebies to Merged column <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened chain <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Commented on demo expectations <sub>or labeled as `no demo`</sub>
- [ ] Pushed merge commit to Gitlab
- [ ] Reindexed `dev` <sub>or this PR does not require reindexing</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Unassign reviewer from PR

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -422,6 +422,24 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
 
 [Terra Data Repository]: https://jade.datarepo-dev.broadinstitute.org/
 
+### 3.2.2 Personal account registration
+
+In order to register your personal email with SAM the following actions need to
+be performed:
+
+Ensure you are logged in with the correct Google Account:
```
Comment (noah-aviel-dove) [inline README.md:430]: Two colons here are very awkward. Based on README precedent, the introductory paragraph should end with a period, and the instruction steps should be numbered.

File: README.md [hunk 2]
```diff
@@ -422,6 +422,24 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
...
+```
+For assistance refer to the Google authorization
+[documentation](https://cloud.google.com/sdk/docs/authorizing).
+
+Register your account with SAM using:
+```
+curl $AZUL_SAM_SERVICE_URL/register/user/v1  -d "" -H 
+"Authorization: Bearer $(gcloud auth --account $account_id print-access-token)"
+```
+
```
Comment (noah-aviel-dove) [inline README.md:442]: Probably a good idea to include a reminder that the account must be then be authorized by an admin after being registered.

**Review by noah-aviel-dove — CHANGES_REQUESTED** [review]
Clear explanation. I think there are a few things missing though.

**Comment by noah-aviel-dove** [comment]
Also, missing check in the checklist.

File: README.md (hunk 1 above)
Comment (amarjandu) [reply README.md:430]: I've omitted the colons for now, perhaps this is something @hannes-ucsc  can clarify with #2169 

Note that we have mixed colon usage for lists of statements for action, some commits to the README add

```
End of action to perform. Run
 Some code block here that helps run that action.

To do some action run:
   Some code block here that helps run that action. 

 To do some action run 
     Some code block here that helps run that action. 

Run 
     Some code block here that helps run that action. 

Then run:
     Some code block here that helps run that action. 
```

The variations of  usage could be due to the age of this README / differences in how we write instructions. Personally I think there should be a guideline to start following so its formatted to look the same.

Interestingly the contributing guidelines use colons before each example code block. While the upgrading.rst has mixed usage.

File: README.md [hunk 3]
```diff
@@ -422,6 +422,30 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
...
+    [documentation](https://cloud.google.com/sdk/docs/authorizing).
+
+2. Register your account with SAM. Run
+
+    ```
+    curl $AZUL_SAM_SERVICE_URL/register/user/v1  -d "" -H 
+    "Authorization: Bearer $(gcloud auth --account $account_id print-access-token)"
+    ```
+
+3. Ask an administrator for TDR to authorize your service account with the
```
Comment (noah-aviel-dove) [inline README.md:446]: This section covers personal accounts, not service accounts.

File: README.md [hunk 4]
```diff
@@ -422,6 +422,30 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
 
 [Terra Data Repository]: https://jade.datarepo-dev.broadinstitute.org/
 
+### 3.2.2 Personal account registration
+Peronsal accounts may want to be registered with SAM to allow access to TDR,
+or for registration as an administrator for TDR. In order to register your 
+personal email with SAM the following actions need to be performed.
```
Comment (noah-aviel-dove) [inline README.md:428]: This is where I need to be a grammar nerd. An account can't "want" anything; only people can have wants or desires. I would reword this to say "Personal accounts may need to be registered"

**Review by noah-aviel-dove — CHANGES_REQUESTED** [review]

**Comment by noah-aviel-dove** [comment]
Also, I'm not sure "administrator for TDR" is the best way of expressing this. "Admin of the 'Stewards' group for the dev environment in Jade" would probably be the most accurate, but I think that's needlessly detailed for this context. We'll see what Hannes has to say.

**Review by noah-aviel-dove — APPROVED** [review]

**Comment by noah-aviel-dove** [comment]
Looks good!

File: README.md [hunk 5]
```diff
@@ -422,6 +422,31 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
 
 [Terra Data Repository]: https://jade.datarepo-dev.broadinstitute.org/
 
+### 3.2.2 Personal account registration
```
Comment (hannes-ucsc) [inline README.md:425]: ambiguous

File: README.md [hunk 6]
```diff
@@ -422,6 +422,31 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
...
+Personal accounts may need to be registered with SAM to allow access to TDR,
```
Comment (hannes-ucsc) [inline README.md:426]: Define

File: README.md [hunk 7]
```diff
@@ -422,6 +422,31 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
...
+or for registration as an administrator within the TDR JadeStewards-dev group.
+In order to register your personal email with SAM the following actions need to
```
Comment (hannes-ucsc) [inline README.md:428]: email or account?

File: README.md [hunk 8]
```diff
@@ -422,6 +422,31 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
...
+be performed.
+
+1. Ensure you are logged in with the correct Google Account. Run
+
+    ```
+    gcloud auth login
+    ```
+
+    For assistance refer to the Google authorization
```
Comment (hannes-ucsc) [inline README.md:437]: documentation can't assist

File: README.md [hunk 9]
```diff
@@ -422,6 +422,31 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
 
 [Terra Data Repository]: https://jade.datarepo-dev.broadinstitute.org/
 
+### 3.2.2 Personal account registration
+Personal accounts may need to be registered with SAM to allow access to TDR,
+or for registration as an administrator within the TDR JadeStewards-dev group.
+In order to register your personal email with SAM the following actions need to
+be performed.
+
+1. Ensure you are logged in with the correct Google Account. Run
```
Comment (hannes-ucsc) [inline README.md:431]: yet another term

File: README.md [hunk 10]
```diff
@@ -422,6 +422,31 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
...
+
+    ```
+    gcloud auth login
+    ```
+
+    For assistance refer to the Google authorization
+    [documentation](https://cloud.google.com/sdk/docs/authorizing).
+
+2. Register your account with SAM. Run
```
Comment (hannes-ucsc) [inline README.md:440]: and another

File: README.md [hunk 11]
```diff
@@ -422,6 +422,31 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
...
+
+    ```
+    curl $AZUL_SAM_SERVICE_URL/register/user/v1  -d "" -H 
+    "Authorization: Bearer $(gcloud auth --account $account_id print-access-token)"
+    ```
+
+3. Ask an administrator of the TDR JadeStewards-dev group to authorize your
```
Comment (hannes-ucsc) [inline README.md:447]: There are three states: absent from the group, present in the group, present and admin of the group

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Remind me to go over this after stand-up.

Regarding colons: English grammar applies. No need to define in the C guide. It's simple:

> Run `ls foo`.

or 

> Run the following command: `ls foo`

In the first variant "`ls foo`" is the subject. In the second, the subject is "the following command". One doesn't need a colon before a subject. Whether the command is formatted inline or as separate block is irrelevant.

**Comment by hannes-ucsc** [comment]
@noah-aviel-dove track review time.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Please discuss this with me after stand-up.

File: README.md [hunk 12]
```diff
@@ -422,6 +422,35 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
 
 [Terra Data Repository]: https://jade.datarepo-dev.broadinstitute.org/
 
+### 3.2.2 Developer Google account registration with SAM
```
Comment (amarjandu) [inline README.md:425]: "SAM registration of Google accounts for developers"

File: README.md [hunk 13]
```diff
@@ -422,6 +422,35 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
...
+
+Developer Google accounts are associated with the developer, this for example is
+your ucsc.edu account. For the remainder of this section it is referred to as 
+your account.
```
Comment (amarjandu) [inline README.md:429]: demarcate against the service account.

Comment (amarjandu) [reply README.md:429]: example after alias.

File: README.md [hunk 14]
```diff
@@ -422,6 +422,35 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
...
+
+Your account may need to be registered with SAM to allow access to TDR,
```
Comment (amarjandu) [inline README.md:431]: s/may/must

Comment (amarjandu) [reply README.md:431]: such that it can access tdr

File: README.md [hunk 15]
```diff
@@ -422,6 +422,35 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
...
+or for registration as an administrator for SAM JadeStewards-dev group. In order
```
Comment (amarjandu) [inline README.md:432]: for you to become an admin of the `JadeStewards-dev` group.

Comment (amarjandu) [reply README.md:432]: the group of Jade stewards (e.g. `JadeStewards-dev`)

File: README.md [hunk 16]
```diff
@@ -422,6 +422,35 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
...
+to register your account with SAM the following actions need to be performed.
+
+1. Ensure you are logged in with the correct account. Run
```
Comment (amarjandu) [inline README.md:435]: your account

Comment (amarjandu) [reply README.md:435]: Login by running

File: README.md [hunk 17]
```diff
@@ -422,6 +422,35 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
...
+
+    ```
+    gcloud auth login
+    ```
+
```
Comment (amarjandu) [inline README.md:440]: When being prompted, select your account.

File: README.md [hunk 18]
```diff
@@ -422,6 +422,35 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
...
+    [documentation](https://cloud.google.com/sdk/docs/authorizing).
+
+2. Register your account with SAM. Run
+
+    ```
+    curl $AZUL_SAM_SERVICE_URL/register/user/v1  -d "" -H 
+    "Authorization: Bearer $(gcloud auth --account $account_id print-access-token)"
+    ```
+
+3. Ask an administrator of the SAM JadeStewards-dev group to add your account
```
Comment (amarjandu) [inline README.md:451]: fix grouping stuff.

File: README.md [hunk 19]
```diff
@@ -422,6 +422,35 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
...
+
+2. Register your account with SAM. Run
+
+    ```
+    curl $AZUL_SAM_SERVICE_URL/register/user/v1  -d "" -H 
+    "Authorization: Bearer $(gcloud auth --account $account_id print-access-token)"
+    ```
+
+3. Ask an administrator of the SAM JadeStewards-dev group to add your account 
+to the group via Slack.
```
Comment (amarjandu) [inline README.md:452]: fix the `via slack` reference

File: README.md [hunk 20]
```diff
@@ -422,6 +422,39 @@ done manually by someone on the Azul team. Ask on Slack to find out more.
...
 [Terra Data Repository]: https://jade.datarepo-dev.broadinstitute.org/
 
+### 3.2.2 SAM registration of Google accounts for developers
+
+While service accounts are associated with Azul, a developer Google account is 
+associated with a developer. For the remainder of this section developer Google 
+accounts will be referred to as your account. For example a developer at UCSC 
+would have an account ending in `ucsc.edu`.
+
+There are two reasons to register your account with SAM, to allow your account
```
Comment (amarjandu) [inline README.md:432]: I changed up the intro to this paragraph in e9e2d9af38ad962dfb5d48a5eeb4331e5185ebe3 . I thought the changes were less awkward.

**Review by hannes-ucsc — APPROVED** [review]
I made more changes, hence the bump.

**Comment by hannes-ucsc** [comment]
For demo, show README section on Github.
